File: services/knowledge-graph/src/knowledge_graph_mcp/tools/sessions.py
"""Exploration session tools — multi-turn context accumulation."""

import json
from typing import Any, Callable, Awaitable
from knowledge_graph_mcp.db import KnowledgeDB


async def _start_session_impl(
    title: str,
    scope: str,
    db: KnowledgeDB,
    user_name: str | None = None,
) -> dict[str, Any]:
    rows = await db.query(
        "INSERT INTO sessions (title, scope, user_name) VALUES ($1, $2, $3) RETURNING id",
        [title, scope, user_name],
    )
    return {"session_id": rows[0]["id"], "scope": scope}


async def _add_context_impl(
    session_id: int,
    entry_type: str,
    content: str,
    db: KnowledgeDB,
    embed_fn: Callable[[str, str, str], Awaitable[list[float]]],
    project_id: str,
    location: str,
    metadata: dict | None = None,
) -> dict[str, Any]:
    seq_rows = await db.query(
        "SELECT MAX(sequence) as max_seq FROM session_entries WHERE session_id = $1",
        [session_id],
    )
    next_seq = (seq_rows[0]["max_seq"] or 0) + 1

    embedding = await embed_fn(content, project_id, location)

    rows = await db.query(
        """INSERT INTO session_entries (session_id, sequence, entry_type, content, metadata, embedding)
           VALUES ($1, $2, $3, $4, $5::jsonb, $6::vector)
           RETURNING id""",
        [session_id, next_seq, entry_type, content, json.dumps(metadata or {}), str(embedding)],
    )

    await db.execute(
        "UPDATE sessions SET updated_at = NOW() WHERE id = $1",
        [session_id],
    )

    return {"entry_id": rows[0]["id"], "sequence": next_seq}


async def _get_session_impl(session_id: int, db: KnowledgeDB) -> dict[str, Any]:
    sessions = await db.query("SELECT * FROM sessions WHERE id = $1", [session_id])
    if not sessions:
        return {"error": f"Session {session_id} not found"}

    entries = await db.query(
        "SELECT id, sequence, entry_type, content, metadata, created_at FROM session_entries WHERE session_id = $1 ORDER BY sequence",
        [session_id],
    )
    return {"session": sessions[0], "entries": entries}


async def _list_sessions_impl(
    db: KnowledgeDB,
    scope: str | None = None,
    status: str | None = None,
    limit: int = 20,
) -> list[dict]:
    conditions = []
    params = []
    param_idx = 1

    if scope:
        conditions.append(f"s.scope = ${param_idx}")
        params.append(scope)
        param_idx += 1

    if status:
        conditions.append(f"s.status = ${param_idx}")
        params.append(status)
        param_idx += 1

    where = f"WHERE {' AND '.join(conditions)}" if conditions else ""
    params.append(limit)

    return await db.query(
        f"""SELECT s.id, s.title, s.scope, s.status, s.user_name,
                   COUNT(e.id) as entry_count, s.created_at, s.updated_at
            FROM sessions s
            LEFT JOIN session_entries e ON e.session_id = s.id
            {where}
            GROUP BY s.id
            ORDER BY s.updated_at DESC
            LIMIT ${param_idx}""",
        params,
    )


def register_session_tools(mcp, get_db, get_settings, embed_fn):
    @mcp.tool()
    async def start_session(title: str, scope: str = "global", user_name: str | None = None) -> str:
        """Create an exploration session — a multi-turn scratchpad for building toward knowledge entries."""
        result = await _start_session_impl(title, scope, get_db(), user_name=user_name)
        return json.dumps(result, default=str)

    @mcp.tool()
    async def add_context(
        session_id: int,
        entry_type: str,
        content: str,
        metadata: dict | None = None,
    ) -> str:
        """Add a context entry to a session. Types: question, observation, code, reference, note."""
        s = get_settings()
        result = await _add_context_impl(
            session_id, entry_type, content, get_db(), embed_fn,
            s.gcp_project_id, s.gcp_location, metadata=metadata,
        )
        return json.dumps(result, default=str)

    @mcp.tool()
    async def get_session(session_id: int) -> str:
        """Load a session's full context history for resumption."""
        result = await _get_session_impl(session_id, get_db())
        return json.dumps(result, default=str)

    @mcp.tool()
    async def list_sessions(scope: str | None = None, status: str | None = None, limit: int = 20) -> str:
        """List recent sessions, optionally filtered by scope or status."""
        result = await _list_sessions_impl(get_db(), scope=scope, status=status, limit=limit)
        return json.dumps(result, default=str)

File: archive/services/knowledge-graph/src/tools/sessions.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { sql, eq, and, desc } from "drizzle-orm";
import type { Db } from "../db/index.js";
import type { Config } from "../config.js";
import { embedText } from "../embeddings.js";
import { sessions, sessionEntries } from "../db/schema.js";

export function registerSessionTools(
  server: McpServer,
  db: Db,
  config: Config,
) {
  server.tool(
    "start_session",
    "Start a new exploration session",
    {
      title: z.string().describe("Session title"),
      scope: z
        .string()
        .optional()
        .default("global")
        .describe("Scope identifier"),
      user_name: z.string().optional().describe("User name"),
    },
    async ({ title, scope, user_name }) => {
      const [session] = await db
        .insert(sessions)
        .values({ title, scope: scope ?? "global", userName: user_name })
        .returning({ id: sessions.id, scope: sessions.scope });

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              session_id: session.id,
              scope: session.scope,
            }),
          },
        ],
      };
    },
  );

  server.tool(
    "add_context",
    "Add a context entry to an exploration session",
    {
      session_id: z.number().describe("Session ID"),
      entry_type: z
        .string()
        .describe("Type: question, observation, code, reference, note"),
      content: z.string().describe("Entry content"),
      metadata: z.record(z.any()).optional().describe("Additional metadata"),
    },
    async ({ session_id, entry_type, content, metadata }) => {
      const [maxSeq] = await db
        .select({
          max: sql<number>`COALESCE(MAX(${sessionEntries.sequence}), 0)`,
        })
        .from(sessionEntries)
        .where(eq(sessionEntries.sessionId, session_id));

      const sequence = (maxSeq?.max ?? 0) + 1;

      const embedding = await embedText(
        content,
        config.gcpProjectId,
        config.gcpLocation,
      );

      const [entry] = await db
        .insert(sessionEntries)
        .values({
          sessionId: session_id,
          sequence,
          entryType: entry_type,
          content,
          embedding,
          metadata: metadata ?? {},
        })
        .returning({
          id: sessionEntries.id,
          sequence: sessionEntries.sequence,
        });

      await db
        .update(sessions)
        .set({ updatedAt: new Date() })
        .where(eq(sessions.id, session_id));

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({
              entry_id: entry.id,
              sequence: entry.sequence,
            }),
          },
        ],
      };
    },
  );

  server.tool(
    "get_session",
    "Get a session with all its context entries",
    {
      session_id: z.number().describe("Session ID"),
    },
    async ({ session_id }) => {
      const [session] = await db
        .select()
        .from(sessions)
        .where(eq(sessions.id, session_id));

      if (!session) {
        return {
          content: [
            {
              type: "text" as const,
              text: JSON.stringify({ error: "Session not found" }),
            },
          ],
        };
      }

      const entries = await db
        .select({
          id: sessionEntries.id,
          sequence: sessionEntries.sequence,
          entryType: sessionEntries.entryType,
          content: sessionEntries.content,
          metadata: sessionEntries.metadata,
          createdAt: sessionEntries.createdAt,
        })
        .from(sessionEntries)
        .where(eq(sessionEntries.sessionId, session_id))
        .orderBy(sessionEntries.sequence);

      return {
        content: [
          {
            type: "text" as const,
            text: JSON.stringify({ session, entries }, null, 2),
          },
        ],
      };
    },
  );

  server.tool(
    "list_sessions",
    "List exploration sessions with optional filters",
    {
      scope: z.string().optional().describe("Filter by scope"),
      status: z.string().optional().describe("Filter by status"),
      limit: z.number().optional().default(20).describe("Max results"),
    },
    async ({ scope, status, limit }) => {
      const conditions: ReturnType<typeof eq>[] = [];
      if (scope) conditions.push(eq(sessions.scope, scope));
      if (status) conditions.push(eq(sessions.status, status));

      const results = await db
        .select({
          id: sessions.id,
          title: sessions.title,
          scope: sessions.scope,
          status: sessions.status,
          userName: sessions.userName,
          createdAt: sessions.createdAt,
          updatedAt: sessions.updatedAt,
          entryCount:
            sql<number>`(SELECT COUNT(*) FROM knowledge_graph.session_entries WHERE session_id = ${sessions.id})`,
        })
        .from(sessions)
        .where(conditions.length > 0 ? and(...conditions) : undefined)
        .orderBy(desc(sessions.updatedAt))
        .limit(limit ?? 20);

      return {
        content: [
          { type: "text" as const, text: JSON.stringify(results, null, 2) },
        ],
      };
    },
  );
}
